src/v2/components/ChannelContents/lib/usePaginatedBlocks.ts

```diff
@@ -0,0 +1,342 @@
+  const getPageFromIndex: UsePaginatedBlocksApi['getPageFromIndex'] = useCallback(
+    index => {
+      return Math.floor(index / channelBlokksPaginatedPerPage) + 1
+    },
+    []
+  )
+
+  /**
+   * Removes a block from a channel ONLY on the frontend. Does not do any
+   * actual mutation/network request.
+   */
+  const removeBlock: UsePaginatedBlocksApi['removeBlock'] = useCallback(
+    ({ id, type }) => {
+      updateCache(({ blockArgs: [prevBlocks, { readField }], prevCount }) => {
+        const newBlocks = prevBlocks.filter(
+          block =>
+            !(
+              readField('id', block) === id &&
+              readField('__typename', block) === type
+            )
+        )
+        return {
+          newBlocks: newBlocks,
+          newCount: prevCount + (newBlocks.length - prevBlocks.length),
+        }
+      })
+    },
+    [updateCache]
+  )
+
+  /**
+   * Moves a block from an old index to a new index and triggers an
+   * apollo mutation
+   */
+  const moveBlock: UsePaginatedBlocksApi['moveBlock'] = useCallback(
+    ({ oldIndex, newIndex }) => {
+      updateCache(({ blockArgs: [prevBlocks, { readField }], prevCount }) => {
+        // Moving to the "bottom". Convert a -1 newIndex value to a
+        // synonymous "count - 1" value that the mutation can understand
+        if (newIndex === -1) {
+          newIndex = prevCount - 1
+        }
+
+        // Get the block reference in the cache. Early exit if we can't
+        // read it
+        const block = prevBlocks[oldIndex]
+        if (!block) {
+          return null
+        }
+
+        // Get the id and typename from the cache. Early exit if we
+        // cant read any of those values
+        const id = readField('id', block) || undefined
+        const typename = readField('__typename', block) || undefined
+        if (id === undefined || typename === undefined) {
+          return null
+        }
+
+        // Fire the mutation
+        client
+          .mutate<
+            moveConnectableMutationData,
+            moveConnectableMutationVariables
+          >({
+            mutation: moveConnectableMutation,
+            variables: {
+              channel_id: args.current.channelId.toString(),
+              connectable: {
+                id: id.toString(),
+                type: getConnectableType(
+                  typename as ChannelContentsConnectable['__typename']
+                ),
+              },
+              insert_at: prevCount - newIndex,
+            },
+          })
+          .then(() => {
+            // After the mutation we need to revalidate pages that have
+            // already been queried, if a block moves into the page
+            // that previously was in a page that wasn't loaded.
+            // This is because we try to not call getPage if hasQueriedPage
+            // returns true, and if an unloaded block moves into the page
+            // then there will be no opportunity for that block to ever load.
```

Comment:
Just pushed up a bugfix I found when moving blocks. If you're in a channel with more than 1 page of blocks and press the "Send to bottom" button on the 1st block, you can get into a state where a block that hasn't been downloaded yet is moved into a page that we are tracking as "already downloaded". The end result is that block is permanently in a loading state until refresh.

This fixes is by querying for all pages in between the block's old position and new position that need to be revalidated.